include/kernels/dslash_clover_helper.cuh

```diff
@@ -184,63 +186,68 @@ namespace quda {
       const int clover_parity = arg.nParity == 2 ? parity : arg.parity;
       const int spinor_parity = arg.nParity == 2 ? parity : 0;
       constexpr int n_flavor = 2;
-      fermion in[n_flavor];
-      fermion out[n_flavor];
 
-#pragma unroll
-      for (int flavor = 0; flavor < n_flavor; flavor++) {
-        in[flavor] = arg.in(x_cb + flavor * arg.volumeCB, spinor_parity);
-        in[flavor].toRel(); // change to chiral basis here
-      }
+      int my_flavor_idx = x_cb + flavor * arg.volumeCB;
+      fermion in = arg.in(my_flavor_idx, spinor_parity);
+      in.toRel(); // change to chiral basis here
 
-#pragma unroll
-      for (int chirality = 0; chirality < 2; chirality++) {
-        // (C + i mu gamma_5 tau_3 - epsilon tau_1 )  [note: appropriate signs carried in arg.a / arg.b]
-        const complex<real> a(0.0, chirality == 0 ? arg.a : -arg.a);
+      int chirality = flavor; // relabel flavor as chirality
+      // (C + i mu gamma_5 tau_3 - epsilon tau_1 )  [note: appropriate signs carried in arg.a / arg.b]
+      const complex<real> a(0.0, chirality == 0 ? arg.a : -arg.a);
 
-        Mat A = arg.clover(x_cb, clover_parity, chirality);
+      Mat A = arg.clover(x_cb, clover_parity, chirality);
 
-        half_fermion in_chi[n_flavor] = {in[0].chiral_project(chirality), in[1].chiral_project(chirality)};
-        half_fermion out_chi[n_flavor];
+      SharedMemoryCache<half_fermion> cache(target::block_dim());
 
+      half_fermion in_chi[n_flavor]; // flavor array of chirally projected fermion
 #pragma unroll
-        for (int flavor = 0; flavor < n_flavor; flavor++) {
-          out_chi[flavor] = A * in_chi[flavor];
-          out_chi[flavor] += (flavor == 0 ? a : -a) * in_chi[flavor];
-          out_chi[flavor] += arg.b * in_chi[1 - flavor];
-        }
+      for (int i = 0; i < n_flavor; i++) in_chi[i] = in.chiral_project(i);
 
-        if (arg.inverse) {
-          if (arg.dynamic_clover) {
-            Mat A2 = A.square();
-            A2 += arg.a2_minus_b2;
-            Cholesky<HMatrix, clover::cholesky_t<real>, N> cholesky(A2);
-#pragma unroll
-            for (int flavor = 0; flavor < n_flavor; flavor++)
-              out_chi[flavor] = static_cast<real>(0.25) * cholesky.backward(cholesky.forward(out_chi[flavor]));
-          } else {
-            Mat Ainv = arg.cloverInv(x_cb, clover_parity, chirality);
-#pragma unroll
-            for (int flavor = 0; flavor < n_flavor; flavor++)
-              out_chi[flavor] = static_cast<real>(2.0) * (Ainv * out_chi[flavor]);
-          }
-        }
+      auto swizzle = [&](half_fermion x[2], int chirality, int reverse) {
+        if (chirality == 0) cache.save_y(x[1], reverse);
+        else                cache.save_y(x[0], 1 - reverse);
+        cache.sync();
+        if (chirality == 0) x[1] = cache.load_y(1 - reverse);
+        else                x[0] = cache.load_y(reverse);
+      };
 
+      swizzle(in_chi, chirality, 0); // apply the flavor-chirality swizzle between threads
```

Comment:
Thoughts on creating a file-scope `enum` that's `FORWARD`/`BACKWARD` or something of the sort that's more explicit than `0` and `1`?

Reply:
acc2681e59785b58ce3e7a50acd2b6f697854540